server/sandbox_list.go

```diff
@@ -8,72 +8,91 @@ import (
-				// and error; it might have been deleted when stop was done.
-				// Log and return an empty struct.
-				log.Warnf(ctx, "Unable to find pod %s with filter", filter.ID)
-				return &types.ListPodSandboxResponse{}, nil
-			}
-			sb := s.getSandbox(id)
-			if sb == nil {
-				podList = []*sandbox.Sandbox{}
-			} else {
-				podList = []*sandbox.Sandbox{sb}
-			}
-		}
-	}
+	podList := s.filterSandboxList(ctx, req.Filter, s.ContainerServer.ListSandboxes())
+	respList := make([]*types.PodSandbox, 0, len(podList))
 
 	for _, sb := range podList {
 		// Skip sandboxes that aren't created yet
 		if !sb.Created() {
 			continue
 		}
 
-		rStatus := types.PodSandboxStateSandboxNotReady
-		if sb.Ready(false) {
-			rStatus = types.PodSandboxStateSandboxReady
-		}
-
 		pod := sb.CRISandbox()
-		pod.State = rStatus
+		pod.State = sb.State()
 
 		// Filter by other criteria such as state and labels.
 		if filterSandbox(pod, req.Filter) {
-			pods = append(pods, pod)
+			respList = append(respList, pod)
 		}
 	}
 
 	return &types.ListPodSandboxResponse{
-		Items: pods,
+		Items: respList,
 	}, nil
 }
+
+// filterSandboxList applies a protobuf-defined filter to retrieve only intended pod sandboxes. Not matching
+// the filter is not considered an error but will return an empty response.
+func (s *Server) filterSandboxList(ctx context.Context, filter *types.PodSandboxFilter, podList []*sandbox.Sandbox) []*sandbox.Sandbox {
+	// Filter by pod id first.
+	if filter == nil {
+		return podList
+	}
+	if filter.ID != "" {
+		id, err := s.PodIDIndex().Get(filter.ID)
+		if err != nil {
+			// Not finding an ID in a filtered list should not be considered
+			// and error; it might have been deleted when stop was done.
+			// Log and return an empty struct.
+			log.Warnf(ctx, "Unable to find pod %s with filter", filter.ID)
+			return []*sandbox.Sandbox{}
+		}
+		sb := s.getSandbox(id)
+		if sb == nil {
+			podList = []*sandbox.Sandbox{}
+		} else {
+			podList = []*sandbox.Sandbox{sb}
+		}
+	}
+	finalList := make([]*sandbox.Sandbox, 0, len(podList))
+	for _, pod := range podList {
+		// Skip sandboxes that aren't created yet
+		if !pod.Created() {
+			continue
+		}
+
+		if filter.State != nil {
+			if pod.State() != filter.State.State {
+				continue
+			}
+		}
+		if filter.LabelSelector != nil {
+			sel := fields.SelectorFromSet(filter.LabelSelector)
+			if !sel.Matches(pod.Labels()) {
+				continue
+			}
+		}
+		finalList = append(finalList, pod)
+	}
+	return finalList
+}
+
+// filterSandbox returns whether passed container matches filtering criteria
+func filterSandbox(p *types.PodSandbox, filter *types.PodSandboxFilter) bool {
+	if filter != nil {
+		if filter.State != nil {
+			if p.State != filter.State.State {
+				return false
+			}
+		}
+		if filter.LabelSelector != nil {
+			sel := fields.SelectorFromSet(filter.LabelSelector)
+			if !sel.Matches(fields.Set(p.Labels)) {
+				return false
+			}
+		}
+	}
+	return true
```

Comment:
Why was this moved? Asking because it makes the review harder (and also git log / git blame).

Reply:
it is a pet peeve of mine when functions are defined before they're called--preventing one reading the file top-down to understand the flow. I figured this one was a relatively small function and not too touch to check for differences